leakpro/user_inputs/cifar10_gia_input_handler.py

```diff
@@ -0,0 +1,82 @@
+"""Module containing the class to handle the user input for the CIFAR10 dataset."""
+
+import logging
+from collections import OrderedDict
+from copy import deepcopy
+
+import torch
+from torch import cuda, device
+from torch.utils.data import DataLoader
+from tqdm import tqdm
+
+from leakpro.fl_utils.gia_optimizers import MetaAdam, MetaOptimizer, MetaSGD
+from leakpro.fl_utils.gia_train import MetaModule
+from leakpro.import_helper import Self
+from leakpro.user_inputs.abstract_gia_input_handler import AbstractGIAInputHandler
+
+
+class Cifar10GIAInputHandler(AbstractGIAInputHandler):
+    """Class to handle the user input for the CIFAR10 dataset."""
+
+    def __init__(self:Self, configs: dict, logger:logging.Logger, client_data: DataLoader, target_model: torch.nn.Module) -> None:
+        super().__init__(configs, logger, client_data, target_model)
+        self.criterion = self.get_criterion()
+
+    def get_criterion(self:Self)->None:
+        """Set the CrossEntropyLoss for the model."""
+        return torch.nn.CrossEntropyLoss()
+
+    def get_optimizer(self: Self) -> MetaOptimizer:
+        """Set the optimizer for the model."""
+        optimizer = self.configs["audit"]["gia_settings"]["optimizer"]
+        lr = self.configs["audit"]["gia_settings"]["learning_rate"]
+        if optimizer == "SGD":
+            return MetaSGD(lr=lr)
+        if optimizer == "Adam":
+            return MetaAdam(lr=lr)
+        raise ValueError(f"Optimizer '{optimizer}' not found. Please check the optimizer settings.")
+
+    def train(
+        self: Self,
+        data: DataLoader = None,
+        optimizer: MetaOptimizer = None,
+    ) -> list:
+        """Model training procedure for GIA.
+
+        This training will create a computational graph through multiple steps, which is necessary
+        for backpropagating to an input image.
+
+        Requires a meta optimizer that performs step to a new set of parameters to keep a functioning
+        graph.
+
+        Training does not update the original model, but returns a norm of what the update would have been.
+        """
+
+        gpu_or_cpu = device("cuda" if cuda.is_available() else "cpu")
+        self.target_model.to(gpu_or_cpu)
+        patched_model = MetaModule(deepcopy(self.target_model))
```

Comment:
do we need to make it a deepcopy?

Reply:
It would be nice if we could make this invisible to the user, i.e., the user shouldnt have to know we have this patched model. Any ideas?